### PR TITLE
fix: prevent tokenizer data leakage during scorer training

### LINKED ISSUE #3035: Improve:
### Is there an existing issue for this?

- [x] I have searched the existing issues

### What would you like to improve?

# Prevent Tokenizer Data Leakage During Scorer Training

## Describe the feature

The scorer training pipeline currently fits the tokenizer on the entire dataset before any train/validation split is performed.

Current implementation:

```python
tokenizer.fit_on_texts(stories + prompts)
```

Since the tokenizer is built using both training and validation samples, information from the validation set is indirectly exposed during training preparation.

## Why this matters

Although the model weights are not trained on validation labels, the vocabulary is still derived from the complete dataset.

This can lead to:

* Data leakage between training and validation sets.
* Overly optimistic evaluation metrics.
* Less realistic measurement of generalization performance.
* Difficulty comparing future experiments using proper ML evaluation practices.

## Proposed Solution

1. Split the dataset into training and validation subsets first.
2. Fit the tokenizer only on the training stories and prompts.
3. Use the fitted tokenizer to transform both training and validation data.
4. Keep validation data completely unseen during vocabulary construction.

Example workflow:

```python
# Split first
X_train, X_val = ...

# Fit tokenizer only on training data
tokenizer.fit_on_texts(train_stories + train_prompts)

# Transform both sets using the same tokenizer
train_encoded = ...
val_encoded = ...
```

## Benefits

* Eliminates vocabulary-level data leakage.
* Produces more reliable validation metrics.
* Aligns the training pipeline with standard machine learning practices.
* Improves experiment reproducibility and model evaluation quality.

## Additional Context

This change does not modify the model architecture or scoring logic. It only improves the correctness of the training and evaluation pipeline.

### Why does this matter?

_No response_

### Screenshots or screen recording (if applicable)

_No response_

### Record

- [x] I agree to follow this project's Code of Conduct
- [x] I want to work on this issue

### PR DESCRIPTION
**Pull request by Ommdutta26**
## Screenshot (when helpful)

N/A – This PR modifies the scorer training pipeline and does not introduce any UI changes.

---

## What this PR does

This PR prevents tokenizer data leakage during scorer model training.

### Changes made

* Split the dataset into training and validation sets before tokenizer fitting.
* Updated tokenizer training to use only training stories and prompts.
* Encoded training and validation data separately using the same tokenizer.
* Replaced `validation_split` with an explicit validation dataset.
* Updated evaluation logic to use the validation set generated by `train_test_split`.

### Why this change is needed

Previously, the tokenizer was fitted on the entire dataset before the train/validation split:

```python
tokenizer.fit_on_texts(stories + prompts)
```

This allowed vocabulary information from the validation set to influence training preparation, resulting in data leakage and potentially optimistic evaluation metrics.

By fitting the tokenizer only on the training data, the validation set remains unseen during vocabulary construction, leading to more reliable model evaluation and better alignment with standard machine learning practices.

---

## Type of change

* [x] Bug fix
* [ ] New feature
* [x] Code refactor
* [ ] Documentation update

---

## Related issue

Closes #3035

---

### CODE REVIEW

**Comment by Ommdutta26** [comment]
@ronisarkarexe kindly review this PR its pending from 4 days
